lib/gmp/gmp.nit

```diff
@@ -0,0 +1,370 @@
+module gmp
+
+private import native_gmp
+
+redef class Numeric
+
+    fun to_l: Long do return self.to_i.to_l
+
+    fun to_r: Ratio do return self.to_f.to_r
+
+end
+
+redef class Text
+
+    #     assert("10".to_l == 10.to_l)
+    fun to_l: Long do
+        var tmp = new NativeMPZ
+        tmp.set_str(self.to_cstring, 10i32)
+        return new Long(tmp)
+    end
+
+    fun to_r: Ratio do
+        var tmp = new NativeMPQ
+        tmp.set_str self.to_cstring
+        return new Ratio(tmp)
+    end
+end
+
+redef class Float
+    redef fun to_l do
+        var tmp = new NativeMPZ
+        tmp.set_d self
+        return new Long(tmp)
+    end
+
+    redef fun to_r do
+        var tmp = new NativeMPQ
+        tmp.set_d self
+        return new Ratio(tmp)
+    end
+end
+
+redef class Int
+    redef fun to_l do
+        var tmp = new NativeMPZ
+        tmp.set_si self
+        return new Long(tmp)
+    end
+
+    redef fun to_r do
+        var tmp = new NativeMPQ
+        tmp.set_si(self, 1)
+        return new Ratio(tmp)
+    end
+end
+
+class Long
+    super Discrete
+    super Numeric
+    super FinalizableOnce
+
+    redef type OTHER: Long
+
+    private var val: NativeMPZ
+
+    redef fun successor(l) do return self + l.to_l
+    redef fun predecessor(l) do return self - l.to_l
+
+    redef fun hash do return self.to_i
+
+    redef fun <=>(l) do
+        if (val.cmp(l.val)) < 0 then
+            return -1
+        else if (val.cmp(l.val)) > 0 then
+            return 1
+        else
+            return 0
+        end
+    end
+
+    redef fun ==(l) do
+        if l isa Long then
+            return (self <=> l) == 0
+        else
+            return false
+        end
+    end
+
+    redef fun <=(l) do return (self <=> l) <= 0
+    redef fun <(l) do return (self <=> l) < 0
+    redef fun >=(l) do return (self <=> l) >= 0
+    redef fun >(l) do return (self <=> l) > 0
+
+
+    #     assert(2.to_l + 2.to_l == 4.to_l)
+    redef fun +(l) do
+        var res = new NativeMPZ
+        val.add(res, l.val)
+        return new Long(res)
+    end
+
+    #     assert(-(2.to_l) == (-2).to_l)
+    redef fun - do
+        var res = new NativeMPZ
+        val.neg res
+        return new Long(res)
+    end
+
+    #     assert(2.to_l - 2.to_l == 0.to_l)
+    redef fun -(l) do
+        var res = new NativeMPZ
+        val.sub(res, l.val)
+        return new Long(res)
+    end
+
+    #     assert(2.to_l * 2.to_l == 4.to_l)
+    redef fun *(l) do
+        var res = new NativeMPZ
+        val.mul(res, l.val)
+        return new Long(res)
+    end
+
+    #     assert(3.to_l / 2.to_l == 1.to_l)
+    redef fun /(l) do
+        var res = new NativeMPZ
+        val.tdiv_q(res, l.val)
+        return new Long(res)
+    end
+
+    #     assert(5.to_l % 2.to_l == 1.to_l)
+    fun %(l: Long): Long do
+        var res = new NativeMPZ
+        val.mod(res, l.val)
+        return new Long(res)
+    end
+
+    #     assert(3.to_l ^ 2 == 9.to_l)
+    fun ^(i: Int): Long do
```

Comment:
I don't know why, but the gmp function of power take a unsigned long as parameter.